event-store-plans.md

```diff
@@ -0,0 +1,174 @@
+# Event Store Improvements for V7
+
+This document is brainstorming for improvements to the Event Store functionality in Marten V7. This does not include features meant for
+"CritterStackPro" products, but does include some necessary baseline infrastructure for planned "CritterStackPro" features.
+
+In some cases, this document outlines "poor man's" subset strategies for features that we will build in a more robust (and inevitably more complicated) way 
+in "CritterStackPro"
+
+The goal is to start a discussion about what is and isn't going to be in scope for V7. When this document is "complete", the output is a new set
+of GitHub issues marked for the V7 release. 
+
+
+## Raise events from asynchronous projection updates
+
+I'm going to allow Oskar address this one:-)
+
+From my side, I think the `ISessionListener` approach we have today is weak and should be a little more "outbox'd" such that it only fires off when
+the transaction is complete. 
+
+For the Wolverine integration, I'd like to be able to enroll in Wolverine's outbox as part of processing a page of events in the daemon to send out messages using the 
+aggregation data as part of the outgoing message.
+
+More generically, we need a good way to emit new events that will be processed as appends within the same daemon transaction. This might be a little tricky and will
+require a spike.
+
+**Gotta do this in a way such that users can opt into whether the new events should be emitted during projection rebuilds**
+
+
+## Sharding the Event Store tables
+
+This is a [long standing issue](https://github.com/JasperFx/marten/issues/770) that we didn't get around to in the huge V4 release.
+
+Some notes:
+
+* My personal proposal is to use native PostgreSQL sharding on the `is_archived` column to effect a "hot storage, cold storage" approach. This should be paired with
+  advice to users to be more aggressive about archiving event streams that are obsolete/stale/not active as a way to improve throughput over all. The daemon and every
+  other place where we query events automatically uses `is_archived = false` filters anyway
+* We probably shouldn't be sharding on anything that would impact the async daemon's "high water detection" that uses the event sequence. In other words, we
+  want to make it as easy as possible to walk the sequence number values within a single table
+* **Maybe** it would be valuable to shard the active events and event streams on the event sequence number ranges. That would be beneficial to the async daemon, if
+  harmful to live aggregations
```

Comment:
Had some thoughts about this, If the streams table was updated to include a "first_event_sequence_id" and "last_event_sequence_id" then these could be used with live projections to drastically cut back on the number of partitions scanned.

Reply:
That's an interesting suggestion. I'll have to think on that one a bit. Might say that's an opt in behavior

Reply:
@elexisvenator could you expand on why would it impact the number of partition?

Reply:
With @elexisvenator 's suggestion, doing a live aggregation of any kind could use the cached "lowest" & "highest" sequence numbers for the stream as part of the WHERE in the event fetching. That would let postgres limit the ranged partitions to look through.

Reply:
it can get costly yes, but a low number of partitions wont make a huge impact. (eg <20, depending on each partitions size). The other challenges with this approach are:

- managing the provisioning of new partitions
- Queries based on timestamps - we can be smarter here with some form of prequery because we know that timestamps increase as sequence id increases but out of the box postgres will have a bad day.
- Unique constraints on event id, these simply cannot exist in any partitioned system, so will affect the partition-on-archived approach as well